src/jrd/Optimizer.cpp

```diff
@@ -1003,26 +1006,35 @@ void OptimizerRetrieval::getInversionCandidates(InversionCandidateList* inversio
 				if (segment->scope == scope)
 					scratch.scopeCandidate = true;
 
-				if (segment->scanType != segmentScanMissing && !(scratch.idx->idx_flags & idx_unique))
+				const USHORT iType = scratch.idx->idx_rpt[j].idx_itype;
+
+				if (iType >= idx_first_intl_string)
 				{
 					const USHORT iType = scratch.idx->idx_rpt[j].idx_itype;
+					TextType* textType = INTL_texttype_lookup(tdbb, INTL_INDEX_TO_TEXT(iType));
 
-					if (iType >= idx_first_intl_string)
+					if (segment->scanType != segmentScanMissing && !(scratch.idx->idx_flags & idx_unique))
 					{
-						TextType* textType = INTL_texttype_lookup(tdbb, INTL_INDEX_TO_TEXT(iType));
-
 						if (textType->getFlags() & TEXTTYPE_SEPARATE_UNIQUE)
 						{
 							// ASF: Order is more precise than equivalence class.
 							// We can't use the next segments, and we'll need to use
 							// INTL_KEY_PARTIAL to construct the last segment's key.
-							scratch.fuzzy = true;
+							scratch.usePartialKey = true;
 						}
 					}
+
+					if (segment->scanType == segmentScanStarting)
+					{
+						if (textType->getFlags() & TEXTTYPE_MULTI_STARTING_KEY)
+							scratch.useMultiStartingKeys = true;	// use INTL_KEY_MULTI_STARTING
+						else
+							scratch.usePartialKey = true;
+					}
```

Comment:
Maybe a better condition sequence would be:
```
if (segment->scanType == segmentScanStarting)
...
else if (segment->scanType != segmentScanMissing && !(scratch.idx->idx_flags & idx_unique))
...
```
?

Currently I see that `scratch.usePartialKey` may be assigned twice.

Reply:
It may be assigned twice, but due to the internal `if's`, the only other way would be to use a single giant `if` with lots of condtions (`(x1 && y1 && z1) || (x2 && y2)`). I think it's going to be more confusing.